src/indexer/updater.ts

```diff
@@ -0,0 +1,459 @@
+    }
+
+    const cap = terms.cap ?? 0n;
+
+    const currentBlockMints = this.mintCounts.get(runeLocation) ?? 0;
+    const totalMints = currentBlockMints + (await this._storage.getValidMintCount(runeLocation));
+
+    if (totalMints >= cap) {
+      return None;
+    }
+
+    const amount = terms.amount ?? 0n;
+
+    this.mintCounts.set(runeLocation, currentBlockMints + 1);
+
+    return Some(amount);
+  }
+
+  private async unallocated(tx: Tx) {
+    const unallocated = new Map<string, u128>();
+
+    for (const input of tx.vin) {
+      if ('coinbase' in input) {
+        continue;
+      }
+
+      const utxoBalance = await this._storage.getUtxoBalance(input.txid, input.vout);
+      for (const additionalBalance of utxoBalance) {
+        const runeLocation = RuneLocation.toString(additionalBalance.runeId);
+        const existingBalance = unallocated.get(runeLocation) ?? u128(0);
+        const newBalance = u128.checkedAddThrow(existingBalance, u128(additionalBalance.amount));
+        unallocated.set(runeLocation, newBalance);
+      }
+    }
+
+    return unallocated;
+  }
+
+  async txCommitsToRune(tx: Tx, rune: Rune): Promise<boolean> {
+    const commitment = rune.commitment;
+    for (const input of tx.vin) {
+      if ('coinbase' in input) {
+        continue;
+      }
+
+      const witnessStack = input.txinwitness.map((item) => Buffer.from(item, 'hex'));
+      const lastWitnessElement = witnessStack[witnessStack.length - 1];
+      const offset =
+        witnessStack.length >= 2 && lastWitnessElement[0] === TAPROOT_ANNEX_PREFIX ? 3 : 2;
+      if (offset > witnessStack.length) {
+        continue;
+      }
+
+      const inscructions = script.decompile(witnessStack[lastWitnessElement.length - offset]);
+      for (const instruction of inscructions) {
+        if (!Buffer.isBuffer(instruction)) {
+          continue;
+        }
+
+        if (Buffer.compare(instruction, commitment) !== 0) {
+          continue;
+        }
+
+        // rpc client
+        const inputTxResult = await this._rpc.getrawtransaction({
+          txid: input.txid,
+          verbose: true,
+        });
+        if (inputTxResult.error !== null) {
+          throw inputTxResult.error;
+        }
+        const inputTx = inputTxResult.result;
+
+        const isTaproot = inputTx.vout[input.vout].scriptPubKey.type === 'witness_v1_taproot';
+        const mature = (inputTx.confirmations ?? -Infinity) >= COMMIT_INTERVAL;
+
+        if (isTaproot && mature) {
+          return true;
+        }
+      }
+    }
+
+    return false;
+  }
+
+  createEtching(txid: string, artifact: Artifact, runeId: RuneLocation, rune: Rune) {
+    if (isRunestone(artifact)) {
+      const { divisibility, terms, premine, spacers, symbol } = artifact.etching.unwrap();
+      this.etchings.push({
+        valid: true,
+        rune: rune.toString(),
+        runeId,
+        txid,
+        ...(divisibility.isSome() ? { divisibility: divisibility.map(Number).unwrap() } : {}),
```

Comment:
Does the monad library have an unwrapOr equivalent?

Reply:
It does, but it doesn't allow for undefined, hence the wonkiness.